File: crates/jobs/src/job.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::id::JobId;
use crate::progress::JobProgress;
use crate::stage::JobStage;
use domain::job::JobStatus;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Job {
    pub id: JobId,
    pub title: String,
    pub project_id: Option<String>,
    pub status: JobStatus,
    pub stage: Option<JobStage>,
    pub progress: JobProgress,
    pub error: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Job {
    pub fn new(title: String, project_id: Option<String>) -> Self {
        let now = Utc::now();
        Self {
            id: JobId::new(),
            title,
            project_id,
            status: JobStatus::Pending,
            stage: None,
            progress: JobProgress::default(),
            error: None,
            created_at: now,
            updated_at: now,
        }
    }
}

File: apps/desktop/src/shared/api/contracts/job.ts
export type JobStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

// Keep future kinds in the global queue; only explicitly mapped kinds affect pipeline steps.
export type JobKind = 'dubbing' | (string & {});

export type JobStage =
  | 'validateSource'
  | 'inspectSubtitles'
  | 'fetchMetadata'
  | 'downloadMedia'
  | 'extractOrGenerateTranscript'
  | 'segmentTranscript'
  | 'translateTranscript'
  | 'prepareDubbingScript'
  | 'synthesizeSegments'
  | 'postprocessAudio'
  | 'muxAudioTrack'
  | 'exportResult';

export type JobProgress = {
  percent: number;
  message: string;
  currentStep: string | null;
  processedItems: number | null;
  totalItems: number | null;
};

export type Job = {
  id: string;
  kind: JobKind;
  revision: number;
  projectId: string | null;
  title: string;
  status: JobStatus;
  stage: JobStage | null;
  progress: JobProgress;
  error: string | null;
  createdAt: string;
  updatedAt: string;
};

export type JobEventKind =
  'created' | 'started' | 'progressed' | 'completed' | 'failed' | 'cancelled';

export type JobEvent = {
  kind: JobEventKind;
  jobId: string;
  revision: number;
  projectId: string | null;
  status: JobStatus;
  stage: JobStage | null;
  progress: JobProgress;
  error: string | null;
};
